gr-analog/lib/sig_source_X_impl.cc.t

```diff
@@ -276,6 +277,13 @@ namespace gr {
       d_ampl = ampl;
     }
 
+    void
+    @NAME@::set_phase(double phase)
+    {
+      d_phase = phase;
+      d_nco.set_phase(phase);
```

Comment:
So, this is a bit problematic: 

GNU Radio is a multithreaded framework, inherently. That means that the block's `work` function might be running the moment someone else calls `set_phase`.

Now, the most common use case is that this block is being used to generate sin or cos waves, and that means that `d_nco` is in a (relatively tight) loop in `fxpt_nco::sincos`, `fxpt_nco::sin` or `fxpt_nco::cos`, where it updates `d_nco.d_phase`. So, this is where we get a race condition, as the thread that called `set_phase` is competing with the thread that runs `work` for write access to `d_nco.d_phase`.

I hear you say "yay! that's not good! So, why aren't the other `set_` methods affected by this?", and the ugly truth is that they *should* take the same precautions, but as frequency and amplitude aren't modified in such tightly-looping manner, the race condition simply doesn't manifest (unless multiple threads call `set_frequency` at once, for example). 

What to do here?

Add a *mutex*! In fact, since this problem is so ubiquitous, the `gr::block` motherclass comes with its own mutex member, `d_setlock`. Using it here would look like:
```c++
   void
   @NAME@::set_phase(double phase)
   {
     gr::thread::scoped_lock l(d_setlock);
     d_phase = phase;
     d_nco.set_phase(phase);
   }
```
…
```c++
int 
@IMPL_NAME@::work ( int noutput_items…) {
 gr::thread::scoped_lock l(d_setlock);
…
}
```
A lock tries to *acquire* a mutex (`d_setlock`, in this case), and blocks for as long as it's currently acquired by another lock. When the `scoped_lock` goes out of scope, it `releases` the mutex. 

So, please, add these two lines that mean the world to me :) (and thread-safety)

Reply:
Added the two mutexes to the work and set_phase methods.